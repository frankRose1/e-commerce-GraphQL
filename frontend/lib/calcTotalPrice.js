/**
 * Take in a users cart and return the total cost of all items in cart
 * @param {array} cart - array of cartItems which contain data about a specific item in the DB
 * @return {integer} totalPrice - total of cart items
 */
function calcTotalPrice(cart){
  return cart.reduce((tally, cartItem) => {
    //in the case an item is deleted from the DB the cartItem will have an ID that points to a deleted item
    if (!cartItem.item) return tally;
    return tally + (cartItem.quantity * cartItem.item.price);
  }, 0);
}

export default calcTotalPrice;