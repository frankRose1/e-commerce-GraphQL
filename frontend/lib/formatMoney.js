/**
 * Format dollar amount of an item
 * @param {integer} amount - dollar amount of an item in cents
 * @return newly formatted price as a string, US currency 
 */
export default function(amount){
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }
  //for whole dollar amounts, leave off the .00
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const formatter = new Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}