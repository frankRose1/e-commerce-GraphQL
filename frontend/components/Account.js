import React, { Component } from 'react';
import gql from 'graphql-tag';
import {Query} from 'react-apollo';
import {formatDistance} from 'date-fns';
import styled from 'styled-components';
import Link from 'next/link';
import formatMoney from '../lib/formatMoney';
import OrderItem from './styles/OrderItemStyles';
import Error from './ErrorMessage';
import Loading from './Loading';
import SubTitle from './styles/SubTitleStyles';

const OrdersGrid = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    userOrders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      items {
        id
        quantity
        description
        title
        price
        image
      }
    }
  }
`;

class Account extends Component {
  render() {
    return (
      <Query query={USER_ORDERS_QUERY}>
        {({data, loading, error}) => {
          const {userOrders} = data;
          if (loading) return <Loading />
          if(error) return <Error error={error} />
          if (!userOrders.length) return <p>Looks like you havn't ordered anything yet!</p>
          
          return (
            <div>
              <SubTitle>Your Account</SubTitle>
              <h2>You have {userOrders.length} order{userOrders.length === 1 ? '' : 's'}</h2>
              <OrdersGrid>
                {userOrders.map(order => (
                  <OrderItem key={order.id}>
                    <Link href={{
                      pathname: '/order',
                      query: { orderId: order.id }
                    }}>
                      <a>
                        <div className="order-meta">
                          <p>{order.items.reduce((a, b) => a + b.quantity, 0)} Items</p>
                          <p>{order.items.length} Product{order.items.length === 1 ? '' : 's'}</p>
                          <p>{formatDistance(order.createdAt, new Date())}</p>
                          <p>{formatMoney(order.total)}</p>
                        </div>
                        <div className="images">
                          {order.items.map(item => (
                            <img 
                              src={item.image} 
                              alt={item.title}
                              key={item.id} />
                          ))}
                        </div>
                      </a>
                    </Link>
                  </OrderItem>
                ))}
            </OrdersGrid>
            </div>
          )
        }}
      </Query>
    );
  }
}

export default Account;