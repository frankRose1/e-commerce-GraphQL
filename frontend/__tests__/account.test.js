import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import Account, { USER_ORDERS_QUERY } from '../components/Account';
import { fakeOrder } from '../lib/testUtils';

const mocks = [
  {
    request: { query: USER_ORDERS_QUERY },
    result: {
      data: {
        userOrders: [
          fakeOrder(),
          {
            ...fakeOrder(),
            id: 'ord124'
          },
          {
            ...fakeOrder(),
            id: 'ord125'
          }
        ]
      }
    }
  }
];

const noOrdersMocks = [
  {
    request: { query: USER_ORDERS_QUERY },
    result: {
      data: {
        userOrders: []
      }
    }
  }
];

describe('<Account />', () => {
  it('renders and matches the snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Account />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('div[data-test="account"]'))).toMatchSnapshot();
  });

  it('displays order information', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Account />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('div[data-test="account"] h2').text()).toContain(
      '3 orders'
    );
  });

  it('renders a message when a user has no orders', async () => {
    const wrapper = mount(
      <MockedProvider mocks={noOrdersMocks}>
        <Account />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('p').text()).toContain(
      "Looks like you havn't ordered anything yet!"
    );
  });
});
