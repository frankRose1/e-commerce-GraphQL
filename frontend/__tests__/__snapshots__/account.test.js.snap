// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account /> renders and matches the snapshot 1`] = `
<div
  data-test="account"
>
  <SubTitleStyles>
    <h4
      className="SubTitleStyles-jg7g5q-0 hcEvTN"
    >
      Your Account
    </h4>
  </SubTitleStyles>
  <h2>
    You have 
    3
     order
    s
  </h2>
  <Account__OrdersGrid>
    <ul
      className="Account__OrdersGrid-ugkfyy-0 hkegQH"
    >
      <OrderItemStyles
        key="ord123"
      >
        <li
          className="OrderItemStyles-sc-3mch2k-0 cErlrh"
        >
          <Link
            href={
              Object {
                "pathname": "/order",
                "query": Object {
                  "orderId": "ord123",
                },
              }
            }
          >
            <a
              href="/order?orderId=ord123"
              onClick={[Function]}
            >
              <div
                className="order-meta"
              >
                <p>
                  2
                   
                  Items
                </p>
                <p>
                  2
                   Product
                  s
                </p>
                <p>
                  8 months
                </p>
                <p>
                  $400
                </p>
              </div>
              <div
                className="images"
              >
                <img
                  alt="soluta non omnis consequatur enim quia autem"
                  key="3a430a73-c9e9-4bcc-b46a-41618965ffea"
                  src="reprehenderit.jpg"
                />
                <img
                  alt="quia sed exercitationem omnis laborum exercitationem est"
                  key="2393e090-592f-4d03-b808-c9d81098deec"
                  src="sint.jpg"
                />
              </div>
            </a>
          </Link>
        </li>
      </OrderItemStyles>
      <OrderItemStyles
        key="ord124"
      >
        <li
          className="OrderItemStyles-sc-3mch2k-0 cErlrh"
        >
          <Link
            href={
              Object {
                "pathname": "/order",
                "query": Object {
                  "orderId": "ord124",
                },
              }
            }
          >
            <a
              href="/order?orderId=ord124"
              onClick={[Function]}
            >
              <div
                className="order-meta"
              >
                <p>
                  2
                   
                  Items
                </p>
                <p>
                  2
                   Product
                  s
                </p>
                <p>
                  8 months
                </p>
                <p>
                  $400
                </p>
              </div>
              <div
                className="images"
              >
                <img
                  alt="nihil quam nulla est unde sit ex"
                  key="b4505a74-1f67-4ee1-9558-e56760116bcd"
                  src="dolorem.jpg"
                />
                <img
                  alt="voluptas culpa et vero repellat eveniet assumenda"
                  key="4cef4042-dadc-4ad4-a066-08d01ced82d4"
                  src="et.jpg"
                />
              </div>
            </a>
          </Link>
        </li>
      </OrderItemStyles>
      <OrderItemStyles
        key="ord125"
      >
        <li
          className="OrderItemStyles-sc-3mch2k-0 cErlrh"
        >
          <Link
            href={
              Object {
                "pathname": "/order",
                "query": Object {
                  "orderId": "ord125",
                },
              }
            }
          >
            <a
              href="/order?orderId=ord125"
              onClick={[Function]}
            >
              <div
                className="order-meta"
              >
                <p>
                  2
                   
                  Items
                </p>
                <p>
                  2
                   Product
                  s
                </p>
                <p>
                  8 months
                </p>
                <p>
                  $400
                </p>
              </div>
              <div
                className="images"
              >
                <img
                  alt="odit tempora at modi occaecati illum laborum"
                  key="1bb8c879-db27-4b15-a8f4-b56bc68196e8"
                  src="at.jpg"
                />
                <img
                  alt="possimus sint fugiat maiores autem est esse"
                  key="dfc616ea-0039-4c01-b135-6d81cafa1cf4"
                  src="et.jpg"
                />
              </div>
            </a>
          </Link>
        </li>
      </OrderItemStyles>
    </ul>
  </Account__OrdersGrid>
</div>
`;
